To keep ubsan happy, Use offsetof() rather than legacy equivalent

RobertHenry6bev opened the pull request:
The existing code causes ubsan to generate a runtime error when using the &((Type *)0->field  idiom.

RobertHenry6bev committed:
To keep ubsan happy, Use offsetof() rather than legacy handmade equivalent.

RobertHenry6bev committed:
Merge branch 'main' of github.com:dotnet/runtime into robhenry-use-offsetof

jkotas commented on src/coreclr/pal/inc/rt/palrt.h:
```diff
@@ -156,21 +156,23 @@ inline void *__cdecl operator new(size_t, void *_P)
+// is conditionally supported.
+// This is more generous than the C++ standards when this code was originally written 
+//
 // PAL_safe_offsetof is a version of offsetof that protects against an
```
You can delete this comment and `PAL_safe_offsetof` and replace all its occurences with just offsetof.

jkotas commented on src/coreclr/pal/inc/rt/palrt.h:
```diff
@@ -156,21 +156,23 @@ inline void *__cdecl operator new(size_t, void *_P)
+// requires offsetof to work correctly even if operator& is overloaded.
+//
 
 #define FIELD_OFFSET(type, field) __builtin_offsetof(type, field)
```
Same for `FIELD_OFFSET`. You can delete the definition and replace the few uses with `offsetof`.

RobertHenry6bev committed:
Merge branch 'main' of github.com:dotnet/runtime into robhenry-use-offsetof

RobertHenry6bev committed:
inline macros PAL_safe_offset and FIELD_OFFSET and replace with offsetof

Using offsetof() will prevent various static analyzers and dynamic ubsan
from complaining about indirecting through NULL.

jkotas commented:
The test failure is https://github.com/dotnet/runtime/issues/73299

jkotas reviewed (approved):
Thank you!
